pkg/clusteragent/appsec/injector.go

```diff
@@ -0,0 +1,324 @@
+		},
+	)
+
+	informerFactory := dynamicinformer.NewFilteredDynamicSharedInformerFactory(si.k8sClient, 0, pattern.Namespace(), func(opts *metav1.ListOptions) {
+		opts.LabelSelector = selector.String()
+	})
+
+	informer := informerFactory.ForResource(pattern.Resource()).Informer()
+
+	handle, err := informer.AddEventHandler(si.createEventHandler(queue))
+	if err != nil {
+		si.logger.Errorf("error adding event handler for resource %s: %s", proxyType, err)
+		return
+	}
+
+	go informer.Run(si.ctx.Done())
+
+	if !cache.WaitForCacheSync(si.ctx.Done(), handle.HasSynced) {
+		si.logger.Warnf("timed out waiting for informer caches to sync for resource %s", proxyType)
+		return
+	}
+
+	health := health.RegisterLiveness("appsec-injector-" + string(proxyType))
+	defer func() {
+		if err := health.Deregister(); err != nil {
+			si.logger.Warnf("error deregistering healthcheck: %s", err)
+		}
+	}()
+
+	si.logger.Debug("Watching resource:", proxyType)
+
+	go func() {
+		for {
+			select {
+			case <-health.C:
+			case <-si.ctx.Done():
+				queue.ShutDown()
+				return
+			}
+		}
+	}()
+
+	for quit := false; !quit; {
+		quit = si.processWorkItem(proxyType, pattern, queue)
+	}
+}
+
+func (si *securityInjector) createEventHandler(queue workqueue.TypedRateLimitingInterface[workItem]) cache.ResourceEventHandler {
+	return cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj any) {
+			unstructured, ok := obj.(*unstructured.Unstructured)
+			if !ok {
+				si.logger.Warnf("event handler for unexpected type: %T", obj)
+			}
+			queue.Add(workItem{name: unstructured.GetName(), namespace: unstructured.GetNamespace(), typ: workItemAdded})
+		},
+		DeleteFunc: func(obj any) {
+			unstructured, ok := obj.(*unstructured.Unstructured)
+			if !ok {
+				si.logger.Warnf("event handler for unexpected type: %T", obj)
+			}
+			queue.Add(workItem{name: unstructured.GetName(), namespace: unstructured.GetNamespace(), typ: workItemDeleted})
+		},
+	}
+}
+
+func (si *securityInjector) processWorkItem(proxyType appsecconfig.ProxyType,
+	pattern appsecconfig.InjectionPattern,
+	queue workqueue.TypedRateLimitingInterface[workItem],
+) bool {
+	item, quit := queue.Get()
+	if quit {
+		return true
+	}
+
+	defer queue.Done(item)
+
+	if !si.isLeader() {
+		// Forget the item to prevent retries when we're not the leader
+		// Only the leader should process proxy configuration changes
+		queue.Forget(item)
+		return false
+	}
+
+	var err error
+	switch item.typ {
+	case workItemAdded:
+		err = pattern.Added(si.ctx, item.namespace, item.name)
+	case workItemDeleted:
+		err = pattern.Deleted(si.ctx, item.namespace, item.name)
+	}
+
+	watchedChangesCounter.Inc(string(proxyType), item.typ.String(), strconv.FormatBool(err == nil))
+
+	if err == nil {
+		queue.Forget(item)
+	} else if queue.NumRequeues(item) < 5 {
+		si.logger.Debugf("requeuing item after error: %v", err)
+		queue.AddRateLimited(item)
+	} else {
+		si.logger.Warnf("unable to process item: %v", err)
```

Comment:
This means that we gave up on processing the item, so I think we need to call Forget so it's removed.